experimental/report/writer.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2020-2025 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package report
+
+import (
+	"bytes"
+	"io"
+	"strings"
+	"unicode"
+
+	"github.com/bufbuild/protocompile/internal/ext/unsafex"
+)
+
+// writer implements low-level writing helpers, including a custom buffering
+// routine to avoid printing trailing whitespace to the output.
+type writer struct {
+	out io.Writer
+	buf []byte
+	err error
+}
+
+// Write implements [io.Writer].
+func (w *writer) Write(data []byte) (int, error) {
+	w.WriteString(unsafex.StringAlias(data))
+	return len(data), nil
+}
+
+func (w *writer) WriteSpaces(n int) {
+	const spaces = "                                        "
+	for n > len(spaces) {
+		w.buf = append(w.buf, spaces...)
+		n -= len(spaces)
+	}
+	w.buf = append(w.buf, spaces[:n]...)
+}
+
+func (w *writer) WriteString(data string) {
+	// Break the input along newlines; each time we're about to append a
+	// newline, discard all trailing whitespace that isn't a newline.
+	for {
+		nl := strings.IndexByte(data, '\n')
+		if nl < 0 {
+			w.buf = append(w.buf, data...)
+			return
+		}
+
+		line := data[:nl]
+		data = data[nl+1:]
+		w.buf = bytes.TrimRightFunc(append(w.buf, line...), unicode.IsSpace)
```

Comment:
We trim in `flush` also... does this mean we don't need to trim here?

Reply:
you're right. :)